database/etcddb/storage.go

```diff
@@ -0,0 +1,205 @@
+package etcddb
+
+import (
+	"github.com/aptly-dev/aptly/database"
+	"github.com/pborman/uuid"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"fmt"
+)
+
+type EtcDStorage struct {
+	url       string
+	db        *clientv3.Client
+	tmpPrefix string // prefix for temporary DBs
+}
+
+// CreateTemporary creates new DB of the same type in temp dir
+func (s *EtcDStorage) CreateTemporary() (database.Storage, error) {
+	tmp := uuid.NewRandom().String()
+	return &EtcDStorage{
+		url:       s.url,
+		db:        s.db,
+		tmpPrefix: tmp,
+	}, nil
+}
+
+func (s *EtcDStorage) applyPrefix(key []byte) []byte {
+	if len(s.tmpPrefix) != 0 {
+		return append([]byte(s.tmpPrefix+"/"), key...)
+	}
+	return key
+}
+
+// Get key value from etcd
+func (s *EtcDStorage) Get(key []byte) (value []byte, err error) {
+	key = s.applyPrefix(key)
+	getResp, err := s.db.Get(Ctx, string(key))
+	if err != nil {
+		return
+	}
+	for _, kv := range getResp.Kvs {
+		value = kv.Value
+		break
+	}
+	if len(value) == 0 {
+		err = database.ErrNotFound
+		return
+	}
+	return
+}
+
+// Put saves key to etcd, if key has the same value in DB already, it is not saved
+func (s *EtcDStorage) Put(key []byte, value []byte) (err error) {
+	key = s.applyPrefix(key)
+	_, err = s.db.Put(Ctx, string(key), string(value))
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Delete removes key from etcd
+func (s *EtcDStorage) Delete(key []byte) (err error) {
+	key = s.applyPrefix(key)
+	_, err = s.db.Delete(Ctx, string(key))
+	if err != nil {
+		return
+	}
+	return
+}
+
+// KeysByPrefix returns all keys that start with prefix
+func (s *EtcDStorage) KeysByPrefix(prefix []byte) [][]byte {
+	prefix = s.applyPrefix(prefix)
+	result := make([][]byte, 0, 20)
+	getResp, err := s.db.Get(Ctx, string(prefix), clientv3.WithPrefix())
+	if err != nil {
+		return nil
+	}
+	for _, ev := range getResp.Kvs {
+		key := ev.Key
+		keyc := make([]byte, len(key))
+		copy(keyc, key)
+		result = append(result, key)
+	}
+	return result
+}
+
+// FetchByPrefix returns all values with keys that start with prefix
+func (s *EtcDStorage) FetchByPrefix(prefix []byte) [][]byte {
+	prefix = s.applyPrefix(prefix)
+	result := make([][]byte, 0, 20)
+	getResp, err := s.db.Get(Ctx, string(prefix), clientv3.WithPrefix())
+	if err != nil {
+		return nil
+	}
+	for _, kv := range getResp.Kvs {
+		valc := make([]byte, len(kv.Value))
+		copy(valc, kv.Value)
+		result = append(result, kv.Value)
+	}
+
+	return result
+}
+
+// HasPrefix checks whether it can find any key with given prefix and returns true if one exists
+func (s *EtcDStorage) HasPrefix(prefix []byte) bool {
+	prefix = s.applyPrefix(prefix)
+	getResp, err := s.db.Get(Ctx, string(prefix), clientv3.WithPrefix())
+	if err != nil {
+		return false
+	}
+	if getResp.Count != 0 {
+		return true
+	}
+	return false
```

Comment:
Could be replaced as `return getResp.Count > 0`

Reply:
done